tests/test_dense.py

```diff
@@ -88,54 +87,83 @@ def test_duplicate_gene(self):
         mock_has_gene_keyword.return_value = False
         mock_header_has_valid_values.return_value = False
+        mock_has_unique_cells.return_value = False
         self.assertFalse(
-            DenseIngestor.is_valid_format(["foo", "foo1"], ["foo2", "foo3"])
+            DenseIngestor.is_valid(
+                ["foo", "foo1"], ["foo2", "foo3"], query_params=query_params
+            )
         )
         self.assertTrue(mock_has_unique_header.called)
         self.assertTrue(mock_has_gene_keyword.called)
         self.assertTrue(mock_header_has_valid_values.called)
+        self.assertTrue(mock_has_unique_cells.called)
 
         # Should raise Value error
         mock_has_unique_header.return_value = True
         mock_has_gene_keyword.return_value = False
         mock_header_has_valid_values.return_value = False
+        mock_has_unique_cells.return_value = False
         self.assertFalse(
-            DenseIngestor.is_valid_format(["foo", "foo1"], ["foo2", "foo3"])
+            DenseIngestor.is_valid(
+                ["foo", "foo1"], ["foo2", "foo3"], query_params=query_params
+            )
         )
 
         # Should raise Value error
         mock_has_unique_header.return_value = True
         mock_has_gene_keyword.return_value = True
+        mock_header_has_valid_values.return_value = True
+        mock_has_unique_cells.return_value = False
+        self.assertFalse(
+            DenseIngestor.is_valid(
+                ["foo", "foo1"], ["foo2", "foo3"], query_params=query_params
+            )
+        )
+        # Should raise Value error
+        mock_has_unique_header.return_value = False
+        mock_has_gene_keyword.return_value = False
         mock_header_has_valid_values.return_value = False
+        mock_has_unique_cells.return_value = True
         self.assertFalse(
-            DenseIngestor.is_valid_format(["foo", "foo1"], ["foo2", "foo3"])
+            DenseIngestor.is_valid(
+                ["foo", "foo1"], ["foo2", "foo3"], query_params=query_params
+            )
         )
 
         # When is_valid_format() returns true
         mock_has_unique_header.return_value = True
         mock_has_gene_keyword.return_value = True
         mock_header_has_valid_values.return_value = True
         self.assertTrue(
-            DenseIngestor.is_valid_format(["foo", "foo1"], ["foo2", "foo3"])
+            DenseIngestor.is_valid(
+                ["foo", "foo1"], ["foo2", "foo3"], query_params=query_params
+            )
         )
 
     @patch("expression_files.expression_files.GeneExpression.load")
     @patch(
         "expression_files.dense_ingestor.DenseIngestor.transform",
         return_value=[("foo1", "foo2")],
     )
-    def test_execute_ingest(self, mock_load, mock_transform):
+    @patch(
+        "expression_files.expression_files.GeneExpression.has_unique_cells",
+        return_value=True,
+    )
+    def test_execute_ingest(self, mock_load, mock_transform, mock_has_unique_cells):
```

Comment:
let's add a unit test where you mock the mongo client .find method, so you can actually test the logic of the has_unique_cells  method, and the error message it generates